eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java

```diff
@@ -60,20 +74,35 @@ public KafkaSubscription subscribe(String subscriberId, Set<String> channels, Ka
     });
   }
 
-  public void handle(RawKafkaMessage message, BiConsumer<Void, Throwable> callback, KafkaMessageHandler kafkaMessageHandler) {
+  private SwimlaneDispatcherBacklog handleBatch(ConsumerRecord<String, byte[]> record,
+                                                SwimlaneBasedDispatcher swimlaneBasedDispatcher,
+                                                BiConsumer<Void, Throwable> callback,
+                                                ReactiveKafkaMessageHandler handler) {
+    return eventuateKafkaMultiMessageConverter
+            .convertBytesToMessages(record.value())
+            .getMessages()
+            .stream()
+            .map(EventuateKafkaMultiMessage::getValue)
+            .map(KafkaMessage::new)
+            .map(kafkaMessage ->
+                    swimlaneBasedDispatcher.dispatch(new RawKafkaMessage(record.value()), record.partition(), message -> handle(message, callback, handler)))
+            .collect(Collectors.toList()) // it is not possible to use "findAny()" now, because streams are lazy and only one message will be processed
+            .stream()
+            .findAny()
+            .get();
+  }
+
+  private void handle(RawKafkaMessage message, BiConsumer<Void, Throwable> callback, ReactiveKafkaMessageHandler kafkaMessageHandler) {
     try {
-      if (eventuateKafkaMultiMessageConverter.isMultiMessage(message.getPayload())) {
-        eventuateKafkaMultiMessageConverter
-                .convertBytesToMessages(message.getPayload())
-                .getMessages()
-                .stream()
-                .map(EventuateKafkaMultiMessage::getValue)
-                .map(KafkaMessage::new)
-                .forEach(kafkaMessageHandler);
-      } else {
-        kafkaMessageHandler.accept(new KafkaMessage(EventuateBinaryMessageEncoding.bytesToString(message.getPayload())));
-      }
-      callback.accept(null, null);
+        kafkaMessageHandler
+                .apply(new KafkaMessage(EventuateBinaryMessageEncoding.bytesToString(message.getPayload())))
+                .thenApply(unused -> {
```

Comment:
Rather than `thenApply()` and `exceptionally()` I think `whenComplete()` would be a better choice

Reply:
ok, I will use it